@vates/nbd-client/index.mjs

```
@@ -0,0 +1,224 @@
+    this.#nbCommands--
+    const magic = await this.#readInt32()
+
+    if (magic !== NBD_REPLY_MAGIC) {
+      throw new Error(`magic number for block answer is wrong : ${magic} ${NBD_REPLY_MAGIC}`)
+    }
+
+    const error = await this.#readInt32()
+    if (error !== 0) {
+      // @todo use error code from constants.mjs
+      throw new Error(`GOT ERROR CODE  : ${error}`)
+    }
+
+    const blockQueryId = await this.#readInt64()
+    const query = this.#commandQueryBacklog.get(blockQueryId)
+    if (!query) {
+      throw new Error(` no query associated with id ${blockQueryId}`)
+    }
+    this.#commandQueryBacklog.delete(blockQueryId)
+    const data = await this.#read(query.size)
+    query.resolve(data)
+
+    this.#waitingForResponse = false
+    if (this.#nbCommands > 0) {
+      await this.#readBlockResponse()
+    }
+  }
+
+  async readBlock(index, size = NBD_DEFAULT_BLOCK_SIZE) {
+    const queryId = this.#nextCommandQueryId
+    this.#nextCommandQueryId++
+
+    // create and send  command at once to ensure there is no concurrency issue
+    const buffer = Buffer.alloc(28)
+    buffer.writeInt32BE(NBD_REQUEST_MAGIC, 0) // it is a nbd request
+    buffer.writeInt16BE(0, 4) // no command flags for a simple block read
+    buffer.writeInt16BE(NBD_CMD_READ, 6) // we want to read a data block
+    buffer.writeBigUInt64BE(queryId, 8)
+    // byte offset in the raw disk
+    buffer.writeBigUInt64BE(BigInt(index) * BigInt(size), 16)
+    buffer.writeInt32BE(size, 24)
+
+    return new Promise(resolve => {
+      // this will handle one block response, but it can be another block
+      // since server does not guaranty to handle query in order
+      this.#commandQueryBacklog.set(queryId, {
+        size,
+        resolve,
+      })
+      // really send the command to the server
+      this.#write(buffer)
+        .then(() => this.#readBlockResponse())
+        .catch(error => {
+          error.data = { index, size }
+          throw error
```

This error is lost, you need to return the promise instead of wrapping it or reject the wrapping promise.

---

I am not sure I can return the promise, since #readBlockResponse() will read one block, but it can be another block